// flow-runtime/lib/__tests__/__fixtures__/ClassDeclaration/typeParametersWithInheritance.js
function _initDefineProp(target, property, descriptor, context) {
  if (!descriptor) return;
  Object.defineProperty(target, property, {
    enumerable: descriptor.enumerable,
    configurable: descriptor.configurable,
    writable: descriptor.writable,
    value: descriptor.initializer ? descriptor.initializer.call(context) : void 0
  });
}

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

function _initializerWarningHelper(descriptor, context) {
  throw new Error('Decorating class property failed. Please ensure that transform-class-properties is enabled.');
}

export function pass(t) {
  var _makeClasses = makeClasses(t),
      MixedPoint = _makeClasses.MixedPoint;

  var point = new MixedPoint([10, true]);
  return point.pass();
}

export function fail(t) {
  var _makeClasses2 = makeClasses(t),
      MixedPoint = _makeClasses2.MixedPoint;

  var point = new MixedPoint([10, true]);
  return point.fail();
}

function makeClasses(t) {
  var _dec, _dec2, _desc, _value, _class, _descriptor, _descriptor2, _class2, _temp, _class3, _temp2;

  var _PointTypeParametersSymbol = Symbol("PointTypeParameters");

  var Point = (_dec = t.decorate(function () {
    return t.flowInto(this[_PointTypeParametersSymbol].T);
  }), _dec2 = t.decorate(function () {
    return t.flowInto(this[_PointTypeParametersSymbol].T);
  }), (_class = (_temp = _class2 = function Point(x, y) {
    babelHelpers.classCallCheck(this, Point);

    _initDefineProp(this, "x", _descriptor, this);

    _initDefineProp(this, "y", _descriptor2, this);

    this[_PointTypeParametersSymbol] = {
      T: t.typeParameter("T")
    };
    var _xType = t.flowInto(this[_PointTypeParametersSymbol].T);
    var _yType = t.flowInto(this[_PointTypeParametersSymbol].T);
    t.param("x", _xType).assert(x);
    t.param("y", _yType).assert(y);
    this.x = x;
    this.y = y;
  }, _class2[t.TypeParametersSymbol] = _PointTypeParametersSymbol, _temp), (_descriptor = _applyDecoratedDescriptor(_class.prototype, "x", [_dec], {
    enumerable: true,
    initializer: function initializer() {
      return 0;
    }
  }), _descriptor2 = _applyDecoratedDescriptor(_class.prototype, "y", [_dec2], {
    enumerable: true,
    initializer: function initializer() {
      return 0;
    }
  })), _class));


  var _MixedPointTypeParametersSymbol = Symbol("MixedPointTypeParameters");

  var MixedPoint = (_temp2 = _class3 = function (_Point) {
    babelHelpers.inherits(MixedPoint, _Point);

    function MixedPoint(_ref) {
      var _ref2 = babelHelpers.slicedToArray(_ref, 2),
          x = _ref2[0],
          y = _ref2[1];

      babelHelpers.classCallCheck(this, MixedPoint);

      var _typeParameters3 = {
        F: t.typeParameter("F")
      };
      t.param("arguments[0]", t.tuple(t.flowInto(_typeParameters3.F), t.flowInto(_typeParameters3.F))).assert(arguments[0]);

      var _this = babelHelpers.possibleConstructorReturn(this, (MixedPoint.__proto__ || Object.getPrototypeOf(MixedPoint)).call(this, x, y));

      _this[_MixedPointTypeParametersSymbol] = _typeParameters3;
      t.bindTypeParameters(_this, _this[_MixedPointTypeParametersSymbol].F);
      return _this;
    }

    babelHelpers.createClass(MixedPoint, [{
      key: "pass",
      value: function pass() {
        var _returnType = t.return(this[_MixedPointTypeParametersSymbol].F);
        _returnType.assert(this.x);
        return _returnType.assert(this.y);
      }
    }, {
      key: "fail",
      value: function fail() {
        var _returnType = t.return(this[_MixedPointTypeParametersSymbol].F);
        return _returnType.assert('bad. this is bad');
      }
    }]);
    return MixedPoint;
  }(Point), _class3[t.TypeParametersSymbol] = _MixedPointTypeParametersSymbol, _temp2);


  return { Point, MixedPoint };
}
