// flow-runtime/lib/types/BooleanType.js
import Type from './Type';
import BooleanLiteralType from './BooleanLiteralType';

import getErrorMessage from "../getErrorMessage";

var BooleanType = function (_Type) {
  babelHelpers.inherits(BooleanType, _Type);

  function BooleanType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, BooleanType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = BooleanType.__proto__ || Object.getPrototypeOf(BooleanType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'BooleanType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(BooleanType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      if (typeof input !== 'boolean') {
        validation.addError(path, this, getErrorMessage('ERR_EXPECT_BOOLEAN'));
        return true;
      }
      return false;
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return typeof input === 'boolean';
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof BooleanType || input instanceof BooleanLiteralType;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return 'boolean';
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName
      };
    }
  }]);
  return BooleanType;
}(Type);

export { BooleanType as default };
