// flow-runtime/lib/types/EmptyType.js
import Type from './Type';

import getErrorMessage from "../getErrorMessage";

var EmptyType = function (_Type) {
  babelHelpers.inherits(EmptyType, _Type);

  function EmptyType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, EmptyType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = EmptyType.__proto__ || Object.getPrototypeOf(EmptyType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'EmptyType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(EmptyType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      validation.addError(path, this, getErrorMessage('ERR_EXPECT_EMPTY'));
      return true;
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return false; // empty types accepts nothing.
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof EmptyType;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return 'empty';
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName
      };
    }
  }]);
  return EmptyType;
}(Type);

export { EmptyType as default };
