// flow-runtime/lib/types/NullLiteralType.js
import Type from './Type';

import getErrorMessage from "../getErrorMessage";

var NullLiteralType = function (_Type) {
  babelHelpers.inherits(NullLiteralType, _Type);

  function NullLiteralType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, NullLiteralType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = NullLiteralType.__proto__ || Object.getPrototypeOf(NullLiteralType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'NullLiteralType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(NullLiteralType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      if (input === null) {
        return false;
      }
      validation.addError(path, this, getErrorMessage('ERR_EXPECT_NULL'));
      return true;
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return input === null;
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof NullLiteralType;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return 'null';
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName
      };
    }
  }]);
  return NullLiteralType;
}(Type);

export { NullLiteralType as default };
