// flow-runtime/lib/types/NumberType.js
import Type from './Type';
import NumericLiteralType from './NumericLiteralType';

import getErrorMessage from "../getErrorMessage";

var NumberType = function (_Type) {
  babelHelpers.inherits(NumberType, _Type);

  function NumberType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, NumberType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = NumberType.__proto__ || Object.getPrototypeOf(NumberType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'NumberType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(NumberType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      if (typeof input === 'number') {
        return false;
      } else {
        validation.addError(path, this, getErrorMessage('ERR_EXPECT_NUMBER'));
        return false;
      }
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return typeof input === 'number';
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof NumberType || input instanceof NumericLiteralType;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return 'number';
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName
      };
    }
  }]);
  return NumberType;
}(Type);

export { NumberType as default };
