// flow-runtime/lib/types/NumericLiteralType.js
import Type from './Type';

import getErrorMessage from "../getErrorMessage";

var NumericLiteralType = function (_Type) {
  babelHelpers.inherits(NumericLiteralType, _Type);

  function NumericLiteralType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, NumericLiteralType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = NumericLiteralType.__proto__ || Object.getPrototypeOf(NumericLiteralType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'NumericLiteralType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(NumericLiteralType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      var value = this.value;

      if (input === value) {
        return false;
      } else {
        validation.addError(path, this, getErrorMessage('ERR_EXPECT_EXACT_VALUE', value));
        return false;
      }
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return input === this.value;
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof NumericLiteralType && input.value === this.value;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return `${ this.value }`;
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName,
        value: this.value
      };
    }
  }]);
  return NumericLiteralType;
}(Type);

export { NumericLiteralType as default };
