// flow-runtime/lib/types/SymbolType.js
import Type from './Type';
import SymbolLiteralType from './SymbolLiteralType';
import getErrorMessage from "../getErrorMessage";

var SymbolType = function (_Type) {
  babelHelpers.inherits(SymbolType, _Type);

  function SymbolType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, SymbolType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = SymbolType.__proto__ || Object.getPrototypeOf(SymbolType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'SymbolType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(SymbolType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      // Issue 252
      if (typeof input === 'symbol') {
        return false;
      } else {
        validation.addError(path, this, getErrorMessage('ERR_EXPECT_SYMBOL'));
        return true;
      }
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return typeof input === 'symbol';
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof SymbolType || input instanceof SymbolLiteralType;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return 'Symbol';
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName
      };
    }
  }]);
  return SymbolType;
}(Type);

export { SymbolType as default };
