// flow-runtime/lib/types/VoidType.js
import Type from './Type';

import getErrorMessage from "../getErrorMessage";

var VoidType = function (_Type) {
  babelHelpers.inherits(VoidType, _Type);

  function VoidType() {
    var _ref;

    var _temp, _this, _ret;

    babelHelpers.classCallCheck(this, VoidType);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = babelHelpers.possibleConstructorReturn(this, (_ref = VoidType.__proto__ || Object.getPrototypeOf(VoidType)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'VoidType', _temp), babelHelpers.possibleConstructorReturn(_this, _ret);
  }

  babelHelpers.createClass(VoidType, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      if (input === undefined) {
        return false;
      } else {
        validation.addError(path, this, getErrorMessage('ERR_EXPECT_VOID'));
        return true;
      }
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      return input === undefined;
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      return input instanceof VoidType;
    }
  }, {
    key: 'toString',
    value: function toString() {
      return 'void';
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName
      };
    }
  }]);
  return VoidType;
}(Type);

export { VoidType as default };
