// flow-runtime/lib/errorReporting/JSONErrorReporter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Validation = require('../Validation');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var JSONErrorReporter = function () {
  function JSONErrorReporter(validation) {
    _classCallCheck(this, JSONErrorReporter);

    this.validation = validation;
  }

  _createClass(JSONErrorReporter, [{
    key: 'report',
    value: function report() {
      var validation = this.validation;

      if (!validation.hasErrors()) {
        return;
      }
      var input = validation.input,
          context = validation.context;

      var errors = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = validation.errors[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var _step$value = _slicedToArray(_step.value, 3),
              path = _step$value[0],
              message = _step$value[1],
              expectedType = _step$value[2];

          var expected = expectedType ? expectedType.toString() : null;
          var actual = context.typeOf((0, _Validation.resolvePath)(input, path)).toString();
          var field = (0, _Validation.stringifyPath)(validation.inputName ? [validation.inputName].concat(path) : path);

          var pointer = '/' + path.join('/');

          errors.push({
            pointer: pointer,
            message: message,
            field: field,
            expected: expected,
            actual: actual
          });
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return { input: input, errors: errors };
    }
  }]);

  return JSONErrorReporter;
}();

exports.default = JSONErrorReporter;
